Spigot-Server-Patches/0553-NamespacedKey-based-plugin-chunk-tickets.patch

```diff
@@ -0,0 +1,267 @@
+index 06e42b7db5385f9a357183552259e7b4491d9fd0..43e5c0695d589e35f23c54f0c18bba0c5e736e79 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+@@ -211,6 +211,39 @@ public class CraftChunk implements Chunk {
+         return getWorld().getPluginChunkTickets(getX(), getZ());
+     }
+ 
++    // Paper start
++    @Override
++    public boolean addKeyedPluginChunkTicket(org.bukkit.NamespacedKey key) {
++        return getWorld().addKeyedPluginChunkTicket(getX(), getZ(), key);
++    }
++
++    @Override
++    public boolean addKeyedPluginChunkTicket(org.bukkit.NamespacedKey key, io.papermc.paper.TickingLevel level) {
++        return getWorld().addKeyedPluginChunkTicket(getX(), getZ(), key, level);
++
++    }
++
++    @Override
++    public boolean removeKeyedPluginChunkTickets(org.bukkit.NamespacedKey key) {
++        return getWorld().removeKeyedPluginChunkTickets(getX(), getZ(), key);
++    }
++
++    @Override
++    public boolean removeKeyedPluginChunkTicket(org.bukkit.NamespacedKey key, io.papermc.paper.TickingLevel level) {
++        return getWorld().removeKeyedPluginChunkTicket(getX(), getZ(), key, level);
++    }
++
++    @Override
++    public Collection<io.papermc.paper.TickingLevel> getKeyedPluginChunkTicketLevels(org.bukkit.NamespacedKey key) {
++        return getWorld().getKeyedPluginChunkTicketLevels(getX(), getZ(), key);
++    }
++
++    @Override
++    public Collection<org.bukkit.NamespacedKey> getKeyedPluginChunkTickets() {
++        return getWorld().getKeyedPluginChunkTickets(getX(), getZ());
++    }
++    // Paper end
++
+     @Override
+     public long getInhabitedTime() {
+         return getHandle().getInhabitedTime();
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index ca0ca5e407661c7f2d1aa3c70990abaeedd816df..4fa5e5d87d54ea54fc6962167009c5e8d7e1a757 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -705,6 +705,132 @@ public class CraftWorld implements World {
+         return ret.entrySet().stream().collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, (entry) -> entry.getValue().build()));
+     }
+ 
++    // Paper start
++    @Override
++    public boolean addKeyedPluginChunkTicket(int x, int z, org.bukkit.NamespacedKey key) {
++        //copy addPluginChunkTicket: ticket level of 31
++        return addKeyedPluginChunkTicket(x, z, key, io.papermc.paper.TickingLevel.ENTITY);
++    }
++
++    @Override
++    public boolean addKeyedPluginChunkTicket(int x, int z, org.bukkit.NamespacedKey key, io.papermc.paper.TickingLevel level) {
++        Preconditions.checkArgument(key != null, "null key");
++        Preconditions.checkArgument(level != null, "null level");
++
++        int ticketLevel = io.papermc.paper.PaperTickingLevel.getTicketLevel(level);
```

Comment:
per other comment, multiple ticking levels per namespaced key is just dirty.

add API should remove any previous tickets associated w/ same key at different levels.
If ticket already exists at that level for this key, should do nothing